tests/data_structures/test_convex_hull_trick.py

```diff
@@ -0,0 +1,46 @@
+from pyrival.data_structures import convex_hull_trick, max_query
+
+def brute(K, M, X):
+    assert(len(K) == len(M))
+
+    out = []
+    for x in X:
+        y = K[0] * x + M[0]
+        for i in range(1, len(K)):
+            y = max(y, K[i] * x + M[i])
+        out.append(y)
+    return out
+
+
+def test_convex_line_hull_integral(n = 1000):
+    import random
+    random.seed(1337)
+
+    K = [random.randint(-1000, 1000) for _ in range(n)]
+    M = [random.randint(-1000, 1000) for _ in range(n)]
+    
+    X = [random.randint(-10000, 10000) for _ in range(n)]
+    brute_ans = brute(K, M, X)
+
+    hull_i, hull_x = convex_hull_trick(K, M)
+    assert(len(hull_i) - 1 == len(hull_x))
+
+    ans = [max_query(x, K, M, hull_i, hull_x) for x in X]
+    assert(ans == brute_ans)
```

Comment:
These tests are really bad. kx+m will be approximately kx when x is large since k ≈ m, so the optimal set of lines will have very small size – for say |x| > 5 (?) it will use the line with largest/smallest k, and only for x near zero will the chosen line be non-trivial. Since only 1000 x's are chosen in [-10000, 10000] that case might never be tested.

For stress-testing algorithms like this I'd strongly recommend running many iterations (10^4–10^5?) where coordinates are tiny (say ≤ 10), only integers are used (so you exercise colinearity and exact equalities), and n varies (between 0 and 20, say).

For testing performance you want artificial cases with more than O(log n) optimal lines, which is what you get from picking random k, m (even if you extend m to be approximately of size x * k).

Reply:
Yeah. I agree the tests were pretty useless. Instead of having one big test case I switched to having 5e4 smaller test cases. Hopefully it looks better now.

As for testing performance. I don't particularly care about testing that here. These tests are just for correctness/testing that the code runs without triggering any exceptions.